docs/4.3. Behaviour - Tokens.md

```diff
@@ -64,113 +62,148 @@ so implementers should take care to ensure they are using the correct epoch.
+the URL for that given API (e.g. "registration", "query", etc.), and the `value` being a JSON object indicating access
+permissions for that API. An empty `x-nmos-api` claim MUST indicate no access to any of the APIs. A missing API key
+MUST indicate no access to that specific API.
+
+#### The Access Permissions Object
+The value of each NMOS API key is the access permissions object for the given user for that specific API. The access
+permission object consists of `key:value` pairs in which the `key` denotes a specific access right. General access
+rights include:
 *   **write** - ability to insert a new, or edit an existing, resource. This takes the form of *POST*, *PUT*,
     *PATCH* and *DELETE* HTTP request verbs.
 *   **read** - ability to retrieve the resource. This takes the form of *GET*, *OPTIONS* and *HEAD* HTTP request
     verbs.
 
-If a value of "write" is given in an access array, and the intention is that the given user also has "read" access to
-a given API, the value of "read" MUST also occur in the array.
+If a specific access right is not permitted for a given user, the access right MUST be omitted, to prevent token bloat.
+If all access rights are not permitted for a given API, the API key MUST be removed from the `x-nmos-api` claim, to
+ensure token size is minimised.
+
+If an access right value of "write" is given in the access permission object, and the intention is that the given user
+also has "read" access to the given API, a value of "read" MUST also be present in the access permissions object.
 
-Individual AMWA specifications MAY specify additional values specific to the individual API. These MUST be registered
-in the [AMWA NMOS Parameter Registers].
+Individual AMWA specifications MAY specify additional access rights specific to the individual API. These MUST be
+registered in the [AMWA NMOS Parameter Registers].
 
-An example of a `x-nmos-api` claim is shown below. This would permit querying an IS-04 Registry, and making
-connections via an IS-05 connection management interface, but would not permit the user to register a resource with
-an IS-04 Registry.
+The `value` corresponding to each access right is a JSON array containing URL path specifiers that the user is
+permitted to perform the specific request against. Wildcards are permitted in the path specifiers. The means by which
+wildcards may be used for pattern matching SHALL align with Section 2.13 of the [POSIX Standard]. For NMOS API
+URL's that follow the standard pattern of:
+
+`<api_proto>://<hostname>:<port>/x-nmos/<api type>/<api version>/`
+
+all path specifiers MUST start _after_ the API version. For example, authorizing a user to read all _single resources_
+and modify the configuration of all  _single senders_ at a Connection Management interface would result in an access
+permissions object of:
+
+```JSON
+"connection": {
+  "read": ["single/*"],
+  "write": ["single/senders/*"]
+}
+```
+
```

Comment:
Having thought about it I think we need a MUST NOT for using anything other than a wildcard where an API specifies a 'URL parameter' (ie. a UUID in IS-04, or a channel label in IS-08). Perhaps in this case a resource server SHOULD always respond with a 401. Otherwise you're going to need handling of the tokens to creep into the code that handles lots of specific API paths. Without this you've got a decent chance of keeping token handling logic in common code away from the handling of specific paths.